src/world/Action/Action.cpp

```diff
@@ -809,11 +819,79 @@ Sapphire::Entity::CharaPtr Action::Action::getHitChara()
 
 bool Action::Action::hasValidLutEntry() const
 {
-  return m_lutEntry.potency != 0 || m_lutEntry.comboPotency != 0 || m_lutEntry.flankPotency != 0 || m_lutEntry.frontPotency != 0 ||
-    m_lutEntry.rearPotency != 0 || m_lutEntry.curePotency != 0 || m_lutEntry.restoreMPPercentage != 0;
+  return m_lutEntry.damagePotency != 0 || m_lutEntry.healPotency != 0 || m_lutEntry.selfHealPotency != 0 || m_lutEntry.selfStatus != 0 ||
+    m_lutEntry.targetStatus != 0 || m_lutEntry.gainMPPercentage != 0;
+}
+
+float Action::Action::getAnimationLock()
+{
+  switch( static_cast< Common::ActionCategory >( m_actionData->actionCategory ) )
+  {
+    case Common::ActionCategory::Item:
+    {
+      return 1.1f;
+    }
+    case Common::ActionCategory::Mount:
+    {
+      return 0.1f;
+    }
+  }
+  return hasCastTime() ? 0.1f : 0.6f;
 }
 
 Action::EffectBuilderPtr Action::Action::getEffectbuilder()
 {
   return m_effectBuilder;
+}
+
+Data::ActionPtr Action::Action::getActionData() const
+{
+  return m_actionData;
+}
+
+Action::ActionEntry Action::Action::getActionEntry() const
+{
+  return m_lutEntry;
+}
+
+void Action::Action::setAutoAttack()
+{
+  m_isAutoAttack = true;
+}
+
+void Action::Action::disableGenericHandler()
+{
+  m_disableGenericHandler = true;
+}
+
+bool Action::Action::isPhysical() const
+{
+  return isAttackTypePhysical( static_cast< Common::AttackType >( m_actionData->attackType ) );
+}
+
+bool Action::Action::isMagical() const
+{
+  return isAttackTypeMagical( static_cast< Common::AttackType >( m_actionData->attackType ) );
+}
+
+bool Action::Action::isAttackTypePhysical( Common::AttackType attackType )
+{
+  return attackType == Common::AttackType::Physical;
+}
+
+bool Action::Action::isAttackTypeMagical( Common::AttackType attackType )
+{
+  return attackType == Common::AttackType::Magical;
+}
+
+void Action::Action::setPrimaryCost( Common::ActionPrimaryCostType type, uint16_t cost )
+{
+  m_primaryCostType = type;
+  m_primaryCost = cost;
+}
+
+Sapphire::StatusEffect::StatusEffectPtr Action::Action::createStatusEffect( uint32_t id, Entity::CharaPtr sourceActor, Entity::CharaPtr targetActor, uint32_t duration, uint32_t tickRate )
+{
+  // workaround to framework access issue in action script
```

Comment:
or just do it properly?

Reply:
the framework in script is a raw pointer